simulator/func_sim/alu.h

```diff
@@ -10,12 +10,84 @@
 #include "multiplication.h"
 #include "traps/trap.h"
 
+#include <array>
 #include <infra/macro.h>
 #include <tuple>
 
 template<size_t N, typename T>
 T align_up(T value) { return ((value + ((1ull << N) - 1)) >> N) << N; }
 
+struct Mask {
+    const uint32 left_mask, right_mask;
+};
+
+static std::array<Mask, 4> shuffle_consts = {
+    Mask{ 0x00ff0000, 0x0000ff00},
+    Mask{ 0x0f000f00, 0x00f000f0},
+    Mask{ 0x30303030, 0x0c0c0c0c},
+    Mask{ 0x44444444, 0x22222222},
+};
+
+static uint32 shuffle_stage( uint32 src, size_t index)
+{
+    uint32 maskL = shuffle_consts[index].left_mask;
+    uint32 maskR = shuffle_consts[index].right_mask;
+    index = 1 << ( 3 - index);
+    uint32 x = src & ~( maskL | maskR);
+    x |= ( ( src << index) & maskL) | ( ( src >> index) & maskR);
+    return x;
+}
+
+static inline uint32 bit_shuffle( uint32 src1, uint32 src2)
+{
+    uint32 result = src1;
+    uint32 shamt = src2 & 15;
+    for( size_t i = 0; i < 4 ; ++i)
+    {
+        if ( shamt & ( 1 << ( 3 - i)))
+            result = shuffle_stage( result, i);
+    }
+    return result;
+}
+
+static inline uint64 bit_shuffle( uint64 src1, uint64 src2)
+{
+    uint64 result = src1;
+    uint32 shamt = src2 & 31;
+    uint32 left_part = result >> 32;
+    uint32 right_part = result;
+    if( shamt & ( 1 << 4))
+    {
+        uint32 buf = left_part & bitmask<uint32>(16);
+        left_part = ( left_part & ~( bitmask<uint32>(16))) | ( ( right_part >> 16) & bitmask<uint32>(16));
+        right_part = ( right_part & ( bitmask<uint32>(16))) | ( buf << 16);
+    }
+    left_part = bit_shuffle( left_part, shamt & 15);
+    right_part = bit_shuffle( right_part, shamt & 15);
+    result = left_part;
+    result <<= 32;
+    return result + right_part;
+}
+
+static inline uint128 bit_shuffle( uint128 src1, uint128 src2)
+{
+    uint128 result = src1;
+    uint64 shamt = src2 & 63;
+    uint64 left_part = result >> 64;
+    uint64 right_part = result;
+    if( shamt & ( 1 << 5))
+    {
+        uint64 buf = left_part & bitmask<uint64>(32);
+        left_part = ( left_part & ~( bitmask<uint64>(32))) | ( ( right_part >> 32) & bitmask<uint64>(32));
+        right_part = ( right_part & ( bitmask<uint64>(32))) | ( buf << 32);
+    }
+    left_part = bit_shuffle( left_part, shamt & 31);
+    right_part = bit_shuffle( right_part, shamt & 31);
+    result = left_part;
+    result <<= 64;
+    return result + right_part;
```

Comment:
If I understand correctly, the only difference against the previous function is 32 vs. 64, right?

Reply:
No, also 15 vs. 31 and 4 vs. 5.

Reply:
But those are 15 = 32 / 2 and log_bitwidth - 1, correct? Then it worth to templatize the function.
The issue with duplicate code is: if someone finds a bug in uint64 version, he or she would need to check uint128.